### PR TITLE
 Add a new consumer that pulls metrics from Google PubSub.

### PR DESCRIPTION
**Pull request by hexedpackets**
In addition to the consumer, Bigtable was updated to 1.4.0, and gRPC was updated to 1.13.1. The shaded names for bigtable dependencies were updated to prevent conflicts with other shaded libraries.

Also upgrading netty and using a single version across all sub-modules. This is to deal with issues arising from attempting to shade multiple versions of netty-tcnative.

### CODE REVIEW

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        log.info("Creating topic {}", topicName);
+        TopicAdminClient topicAdminClient = TopicAdminClient.create(
+            TopicAdminSettings.newBuilder()
+                .setTransportChannelProvider(channelProvider)
+                .setCredentialsProvider(credentialsProvider)
+                .build()
+        );
+        try {
+            topicAdminClient.createTopic(topicName);
+        } catch (AlreadyExistsException e) {
```
Comment (jsferrei) [inline consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java]: I assume this is the only exception that is OK, right?

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+import lombok.NoArgsConstructor;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@Data
+public class PubSubConsumerModule implements ConsumerModule {
+    private static final int DEFAULT_THREADS_PER_SUBSCRIPTION = 8;
+    private static final Long DEFAULT_MAX_OUTSTANDING_ELEMENT_COUNT = 10_000L;
+    private static final Long DEFAULT_MAX_OUTSTANDING_REQUEST_BYTES = 1_000_000_000L;
```
Comment (jsferrei) [inline consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java]: Again, would we want any of these configurable?

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+            @Named("consuming") AtomicInteger consuming,
+            @Named("total") AtomicInteger total,
+            @Named("errors") AtomicLong errors,
+            @Named("consumed") LongAdder consumed
+        ) {
+            return async.managed(new ManagedSetup<Connection>() {
+                @Override
+                public AsyncFuture<Connection> construct() {
+                    return async.call(() -> {
+                        log.info("p:{}, t:{}, s:{}", projectId, topicId, subscriptionId);
```
Comment (jsferrei) [inline consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java]: Just for clarity, can we change the log text to read "projectId: {projectId}, topicId:{topicId}, subscriptionId:{subscriptionId}"

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+                        connection.createSubscription();
+
+                        return connection.start();
+                    });
+                }
+
+                @Override
+                public AsyncFuture<Void> destruct(final Connection connection) {
+                    connection.shutdown();
+                    total.set(0);
```
Comment (jsferrei) [inline consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java]: Do we need to reset any of the other values? Consuming, errors, consumed, etc

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+               long maxOutstandingRequestBytes
+    ) {
+        this.consumer = consumer;
+        this.projectId = projectId;
+        this.subscriptionId = subscriptionId;
+        this.threads = threads;
+        this.topicName = ProjectTopicName.of(projectId, topicId);
+        this.subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId);
+        this.credentialsProvider = SubscriptionAdminSettings.defaultCredentialsProviderBuilder().build();
+        this.channelProvider = SubscriptionAdminSettings
```
[the remark this answers is not in the document]
Comment (hexedpackets) [reply consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java]: Yes, I'll update them.

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java (hunk 4 above)
Comment (hexedpackets) [reply consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java]: I haven't read through the code that uses those so I'm not completely certain, but the kafka consumer just sets total to 0 in its destruct method.

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java (hunk 3 above)
Comment (hexedpackets) [reply consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java]: :+1:

Comment (hexedpackets) [reply consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java]: That doesn't seem to work with slf4j.

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java [hunk 6]
```diff
@@ -53,7 +53,7 @@
     private final ConsumerSchema.Consumer consumer;
     private final long maxOutstandingElementCount;
     private final long maxOutstandingRequestBytes;
-    private Subscriber subscriber = null;
```
Comment (dmichel1) [inline consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java]: can you make this `final` now to remain consistent with the other variables?

File: heroic-component/src/main/java/com/spotify/heroic/metric/Metric.java [hunk 7]
```diff
@@ -40,7 +40,7 @@ static Metric invalid() {
         return invalid;
     }
 
-    Comparator<Metric> comparator = (a, b) -> Long.compare(a.getTimestamp(), b.getTimestamp());
+    Comparator<Metric> comparator = Comparator.comparingLong(Metric::getTimestamp);
```
Comment (dmichel1) [inline heroic-component/src/main/java/com/spotify/heroic/metric/Metric.java]: What's this change about, it seems unrelated(?)

File: heroic-dist/src/test/java/com/spotify/heroic/PubSubConsumerIT.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import net.jcip.annotations.NotThreadSafe;
+import org.junit.After;
+
+@NotThreadSafe
+public class PubSubConsumerIT extends AbstractConsumerIT {
+    private final String topic = "topic1";
+    private final String subscription = "sub1";
+    private final String project = "xpn-heroic-it-1";
```
Comment (dmichel1) [inline heroic-dist/src/test/java/com/spotify/heroic/PubSubConsumerIT.java]: want to drop the `xpn` bit here?

File: heroic-component/src/main/java/com/spotify/heroic/metric/Metric.java (hunk 7 above)
Comment (hexedpackets) [reply heroic-component/src/main/java/com/spotify/heroic/metric/Metric.java]: I was looking at how metrics get ordered (needed to sort them in the IT) and updated this. Its the exact same logic just less verbose.

File: heroic-dist/src/test/java/com/spotify/heroic/PubSubConsumerIT.java (hunk 8 above)
Comment (hexedpackets) [reply heroic-dist/src/test/java/com/spotify/heroic/PubSubConsumerIT.java]: :+1:

File: consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java (hunk 6 above)
Comment (hexedpackets) [reply consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java]: :+1:

Comment (hexedpackets) [reply consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/Connection.java]: It can't be final since it isn't set in the constructor.
